src/dune/toplevel.ml

```diff
@@ -59,9 +62,39 @@ end
 type t =
   { cctx : Compilation_context.t
   ; source : Source.t
+  ; preprocess : Dune_file.Preprocess.t
   }
 
-let make ~cctx ~source = { cctx; source }
+let make ~cctx ~source ~preprocess = { cctx; source; preprocess }
+
+let pp_flags t =
+  let open Pp.O in
+  let sctx = Compilation_context.super_context t.cctx in
+  let scope = Compilation_context.scope t.cctx in
+  let expander = Compilation_context.expander t.cctx in
+  match t.preprocess with
+  | Pps { loc; pps; flags; staged = _ } -> (
+    match
+      Preprocessing.get_ppx_driver sctx ~loc ~expander
+      ~lib_name:None ~flags ~scope pps
+    with
+    | Error _exn -> Pp.nop
+    | Ok (exe, flags) ->
+      let ppx =
+        Dyn.Encoder.list
+          Dyn.Encoder.string
+            [(Path.to_absolute_filename (Path.build exe)
+             :: "--as-ppx" :: flags
+             |> String.concat ~sep:" ")]
+      in
+      (* Set Clflags.all_ppx for dune utop, and Compenv.first_ppx for custom
+         toplevels because Topmain.main() resets Clflags.all_ppx. *)
+      Pp.vbox ~indent:2
+        (Pp.verbatim "Clflags.all_ppx :=" ++ Pp.cut ++ Dyn.pp ppx)
+      ++ Pp.verbatim ";" ++ Pp.newline
+      ++ Pp.verbatim "Compenv.first_ppx :=" ++ Pp.cut ++ Dyn.pp ppx
+      ++ Pp.verbatim ";" ++ Pp.newline)
+  | Action _ | Future_syntax _ | No_preprocessing -> Pp.nop
```

Comment:
If we don't support `Action _` or `Future_syntax _` for now, perhaps we should error out instead.